src/runtime/gc_blocks.go

```diff
@@ -700,45 +674,40 @@ func markRoot(addr, root uintptr) {
 }
 
 // Sweep goes through all memory and frees unmarked memory.
-// It returns how many bytes are free in the heap after the sweep.
-func sweep() (freeBytes uintptr) {
-	freeCurrentObject := false
-	var freed uint64
-	for block := gcBlock(0); block < endBlock; block++ {
-		switch block.state() {
-		case blockStateHead:
-			// Unmarked head. Free it, including all tail blocks following it.
-			block.markFree()
-			freeCurrentObject = true
-			gcFrees++
-			freed++
-		case blockStateTail:
-			if freeCurrentObject {
-				// This is a tail object following an unmarked head.
-				// Free it now.
-				block.markFree()
-				freed++
-			}
-		case blockStateMark:
-			// This is a marked object. The next tail blocks must not be freed,
-			// but the mark bit must be removed so the next GC cycle will
-			// collect this object if it is unreferenced then.
-			block.unmark()
-			freeCurrentObject = false
-		case blockStateFree:
-			freeBytes += bytesPerBlock
-		}
+func sweep() {
+	endBlock := endBlock
+	metadataEnd := unsafe.Add(metadataStart, (endBlock+(blocksPerStateByte-1))/blocksPerStateByte)
+	var carry byte
+	for meta := metadataStart; meta != metadataEnd; meta = unsafe.Add(meta, 1) {
+		// Fetch the state byte.
+		stateBytePtr := (*byte)(unsafe.Pointer(meta))
+		stateByte := *stateBytePtr
+
+		// Separate blocks by type.
+		high := stateByte >> blocksPerStateByte
+		markedHeads := stateByte & high
+		unmarkedHeads := (stateByte & blockStateEach) &^ high
+		tails := high &^ stateByte
+
+		// Adding 1 to a run of bits will clear the run.
+		// Add 1 to the next bit in the tails mask after a freed head to clear the corresponding tails.
+		// Carry the overflow between state bytes.
+		tailClear := tails + (unmarkedHeads << 1) + carry
```

Comment:
Example here please.  Maybe put some of these bit manips into functions which could have their own unit tests?